server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -392,20 +392,35 @@ public Integer readOptionalVInt() throws IOException {
         return null;
     }
 
-    // we don't use a CharsRefBuilder since we exactly know the size of the character array up front
+    // Maximum char-count to de-serialize via the thread-local CharsRef buffer
+    private static final int SMALL_STRING_LIMIT = 1024;
+
+    // Thread-local buffer for smaller strings
+    private static final ThreadLocal<CharsRef> smallSpare = ThreadLocal.withInitial(() -> new CharsRef(SMALL_STRING_LIMIT));
+
+    // Larger buffer used for long strings that can't fit into the thread-local buffer
+    // We don't use a CharsRefBuilder since we exactly know the size of the character array up front
     // this prevents calling grow for every character since we don't need this
-    private final CharsRef spare = new CharsRef();
+    private CharsRef largeSpare;
```

Comment:
Can this be a SetOnce?

Reply:
I'd rather not do that. Then we'd still needlessly create `SetOnce` (which seems about the same size of an empty `CharsRef`) here and even add a `volatile` read into the process.